### PR TITLE
Move MySQL packet parsing to individual functions

### PR DESCRIPTION
**Pull request by greedy52**
A small follow up PR of
- https://github.com/gravitational/teleport/pull/10283

No parsing logic change. Just rearranging things to have one parser function per packet type, and use a slice to replace the switch for tiny performance improvement. UT all passed.

### CODE REVIEW

File: lib/srv/db/mysql/protocol/packet.go [hunk 1]
```diff
@@ -108,103 +53,13 @@ func ParsePacket(conn io.Reader) (Packet, error) {
...
-		}
-		return packet, nil
-
-	case packetTypeStatementBulkExecute:
-		packet, ok := parseStatementBulkExecutePacket(packet)
-		if !ok {
-			return nil, trace.BadParameter("failed to parse COM_STMT_BULK_EXECUTE packet: %v", packetBytes)
-		}
-		return packet, nil
+	if int(packetType) < len(packetParsersByType) && packetParsersByType[packetType] != nil {
```
Comment (Tener) [inline lib/srv/db/mysql/protocol/packet.go:56]: This is an interesting approach to "use a slice to replace the switch for tiny performance improvement" (quoting the PR description). Naively I'd expect the switch to have better performance. Are the benefits of this approach discussed anywhere?

Comment (greedy52) [reply lib/srv/db/mysql/protocol/packet.go:56]: my believe is switch is `o(n)` but slice is `o(1)`. i couldn't a good post ppl comparing slice vs switch. switch is mostly compared with map. here is one stackoverflow has some benchmark https://stackoverflow.com/questions/46789259/map-vs-switch-performance-in-go. Though I believe switch will get slower when n gets large.

I will do some benchmark code for fun and share

Comment (greedy52) [reply lib/srv/db/mysql/protocol/packet.go:56]: some rough numbers
```
go test -bench=. -benchtime=10s

goos: darwin
goarch: arm64
pkg: github.com/greedy52/golang-switch-slice-map-compare/intkey
BenchmarkBaseline-8   	860302086	        13.52 ns/op
BenchmarkSwitch1-8    	888747649	        13.50 ns/op
BenchmarkSwitch5-8    	828560742	        14.49 ns/op
BenchmarkSwitch10-8   	489604197	        24.49 ns/op
BenchmarkSwitch20-8   	398235208	        30.13 ns/op
BenchmarkSlice-8      	887899017	        13.49 ns/op
BenchmarkMap-8        	339221132	        35.84 ns/op
```

https://github.com/greedy52/golang-switch-slice-map-compare/blob/main/intkey/benchmark_test.go

**Review by Tener — APPROVED** [review]
LGTM.

**Review by gabrielcorado — APPROVED** [review]

File: lib/srv/db/mysql/protocol/packet.go [hunk 2]
```diff
@@ -108,103 +53,13 @@ func ParsePacket(conn io.Reader) (Packet, error) {
...
+		return packetParsersByType[packetType](packetBytes)
```
Comment (smallinsky) [inline lib/srv/db/mysql/protocol/packet.go:57]: nit: Can we also propagate this call in stack-trace ? 
```suggestion
		packet, err packetParsersByType[packetType](packetBytes)
		if err != nil {
		   return nil, trace.Wrap(err)
		}
		return packet, nil
```

**Review by smallinsky — APPROVED** [review]

**Pull request merged**